kolibri/content/serializers.py

```diff
@@ -84,52 +85,46 @@ class ContentNodeSerializer(serializers.ModelSerializer):
         lookup_field_2='pk',
     )
 
+    # Here we use a FileSerialize instead just the files reverse FK is because we want to get the computed field storage_url
+    # In order to improve performance in production, we should implement a client side method to calculate the file url using
+    # extension and checksum field along with the setting.STORAGE_ROOT, which can be passed to front end at template boostrapping.
     files = FileSerializer(many=True, read_only=True)
-    ancestor_ids = serializers.SerializerMethodField()
-    immediate_children_ids = serializers.SerializerMethodField()
-    preload = serializers.SerializerMethodField()
-
-    def get_ancestor_ids(self, target_node):
+    ancestors = serializers.SerializerMethodField()
+
+    def __init__(self, *args, **kwargs):
+        # Instantiate the superclass normally
+        super(ContentNodeSerializer, self).__init__(*args, **kwargs)
+
+        # enable dynamic fields specification!
+        if not isinstance(self.context['request'], WSGIRequest):
+            if 'request' in self.context and self.context['request'].query_params.get('fields'):
+                fields = self.context['request'].query_params.get('fields').split(',')
+                # Drop any fields that are not specified in the `fields` argument.
+                allowed = set(fields)
+                existing = set(self.fields.keys())
+                for field_name in existing - allowed:
+                    self.fields.pop(field_name)
```

Comment:
self.fields is not a tuple. The data structure looks like this:
```
{'files': FileSerializer(many=True, read_only=True):
    url = DualLookuplinkedIdentityField(lookup_field_1='channelmetadata_channel_id', lookup_field_2='pk', view_name='file-detail')
    storage_url = SerializerMethodField()
    id = IntegerField(label='ID', read_only=True)
    priority = IntegerField(allow_null=True, required=False)
    checksum = CharField(allow_blank=True, max_length=400, required=False)
    available = BooleanField(required=False)
    file_size = IntegerField(allow_null=True, required=False)
    ...
```
I'm not sure what this data structure is. but pop works on it.
Also self.fields is different from fields, existing and allowed. 
self.fields = existing.intersection(allowed) will not work.